posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,172 @@
+Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.
+
+## How simp works
+
+On an informal level, `simp` can be seen as recursively traversing the expression `e` to be simplified from the outside-in, each time checking whether the expression matches the left hand side `LHS` of some simp lemma of the form `LHS = RHS`, and replacing `LHS` by `RHS` in the expression if it does match.
+
+For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+> To see what proof steps `simp` used, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.
+
+In this picture, simp lemmas are *fixed* rules to turn a left hand side of a *specific* shape into a right hand side of a *specific* shape.
+In contrast, simprocs are *flexible* rules to turn a left hand side of a *specific* shape into a right hand side of a shape *computed* from the left hand side.
+In this sense, they are *parametric simp lemmas*.
+
+## Examples of simprocs
+
+In this section, we exemplify three simprocs through the following use cases:
+* Computation
+* Avoiding combinatorial explosion of lemmas
+* Performance optimisation
+
+### Computation:
+
+#### The `Finset.Icc_ofNat_ofNat` simproc
+
+The `Finset.Icc_ofNat_ofNat` simproc is designed to take expressions of the form `Finset.Icc a b` where `a` and `b` are numerals, and simplify them to an explicit set.
+
+```lean
+example : Finset.Icc 1 0 = ∅ := by
+  simp only [Icc_ofNat_ofNat]
+
+example : Finset.Icc 1 1 = {1} := by
+  simp only [Icc_ofNat_ofNat]
+
+example : Finset.Icc 1 4 = {1, 2, 3, 4} := by
+  simp only [Icc_ofNat_ofNat]
+
+example : Finset.Icc a (a + 2) = {a, a + 1, a + 2} := by
+  -- fails: the bounds of the interval aren't numerals!
+  simp only [Icc_ofNat_ofNat]
+```
+
+#### The `Nat.reduceDvd` simproc
+
+The `Nat.reduceDvd` simproc is designed to take expressions of the form `a | b` where `a, b` are natural number literals, and simplify them to `True` or `False`.
+
+```lean
+example : 3 ∣ 9 := by
+  simp only [Nat.reduceDvd]
+
+example : ¬ 2 ∣ 49 := by
+  simp only [Nat.reduceDvd]
+  -- Remaining goal: `¬ False`
+  simp
+
+example : a ∣ a * b := by
+  simp only [Nat.reduceDvd] --fails
+```
+
+Here the metaprogram run by `Nat.reduceDvd` does the following whenever an expression of the form `a ∣ b` (where `a, b` are natural numbers) is encountered:
+- Check that `∣` is the usual natural number division.
+- Check that `a` and `b` are numerals.
+- Compute `b % a`.
+- If this quantity is zero, then return `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.
+- If this quantity isn't zero, then return `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.
```

Comment:
I am not sure how I feel about these technicalish explanations being so early in the blog post. Should they be relegated outside of `## Examples`? to the second blog post? I am open to suggestions

Reply:
I'm pretty content with this as it is: you mention the right hand side is computed from the left, and giving an overview of how that computation could work is helpful imo. I don't see this as too technical (although you probably could omit the part about checking `∣`)

Reply:
I agree - this is something I meant to remove (e.g. "Check that `∣` is the usual natural number division." is a very confusing sentence if you can't see the corresponding code)